{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63dcf077-f6c8-404a-8fc3-96784d0fd87a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# python 環境變數\n",
    "import os\n",
    "HOSTNAME=os.environ['HOSTNAME']\n",
    "os.environ['http_proxy'] = \"socks5:/\"+HOSTNAME+\":12345\" \n",
    "os.environ['https_proxy'] = \"socks5://\"+HOSTNAME+\":12345\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8e92169-b557-4962-a244-08ec85f21c87",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "IMAGE=/work1/c00cjz00/f1/docker/humann_latest.sif\n",
    "IMAGE_basename=$(basename \"$IMAGE\" .sif)\n",
    "IPYKERNEL=/work1/c00cjz00/f1/ipykernel/image_without_ipykernel\n",
    "mkdir -p /home/$(whoami)/.local/share/jupyter/kernels/\n",
    "rm -rf /home/$(whoami)/.local/share/jupyter/kernels/${IMAGE_basename}\n",
    "sleep 3\n",
    "cp -rf ${IPYKERNEL} /home/$(whoami)/.local/share/jupyter/kernels/${IMAGE_basename}\n",
    "chmod -R 755 /home/$(whoami)/.local/share/jupyter/kernels/${IMAGE_basename}\n",
    "IMAGE_desc=$(echo $IMAGE | sed 's_/_\\\\/_g')\n",
    "sed -i \"s/templateSIF/${IMAGE_desc}/g\" /home/$(whoami)/.local/share/jupyter/kernels/${IMAGE_basename}/kernel.json\n",
    "sed -i \"s/templateImage/Image_${IMAGE_basename}/g\" /home/$(whoami)/.local/share/jupyter/kernels/${IMAGE_basename}/kernel.json\n",
    "singularity exec -B /pkg/tools/miniconda3 ${IMAGE} /pkg/tools/miniconda3/bin/pip install ipykernel -q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2ddbc74a-a231-4ca3-bbae-e453aefeb96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#初始環境設定\n",
    "import os\n",
    "from pathlib import Path\n",
    "HOME = str(Path.home())\n",
    "Add_Binarry_Path=HOME+'/.local/bin:/usr/localbin'\n",
    "os.environ['PATH']=os.environ['PATH']+':'+Add_Binarry_Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92464249-1365-48b9-a685-7d9745355d9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: humann [-h] -i <input.fastq> -o <output> [--threads <1>] [--version]\n",
      "              [-r] [--bypass-nucleotide-index] [--bypass-nucleotide-search]\n",
      "              [--bypass-prescreen] [--bypass-translated-search]\n",
      "              [--taxonomic-profile <taxonomic_profile.tsv>]\n",
      "              [--memory-use {minimum,maximum}]\n",
      "              [--input-format {fastq,fastq.gz,fasta,fasta.gz,sam,bam,blastm8,genetable,biom}]\n",
      "              [--search-mode {uniref50,uniref90}] [-v]\n",
      "              [--metaphlan <metaphlan>]\n",
      "              [--metaphlan-options <metaphlan_options>]\n",
      "              [--prescreen-threshold <0.01>] [--bowtie2 <bowtie2>]\n",
      "              [--bowtie-options <bowtie_options>]\n",
      "              [--nucleotide-database <nucleotide_database>]\n",
      "              [--nucleotide-identity-threshold <0.0>]\n",
      "              [--nucleotide-query-coverage-threshold <90.0>]\n",
      "              [--nucleotide-subject-coverage-threshold <50.0>]\n",
      "              [--diamond <diamond>] [--diamond-options <diamond_options>]\n",
      "              [--evalue <1.0>] [--protein-database <protein_database>]\n",
      "              [--rapsearch <rapsearch>]\n",
      "              [--translated-alignment {usearch,rapsearch,diamond}]\n",
      "              [--translated-identity-threshold <Automatically: 50.0 or 80.0, Custom: 0.0-100.0>]\n",
      "              [--translated-query-coverage-threshold <90.0>]\n",
      "              [--translated-subject-coverage-threshold <50.0>]\n",
      "              [--usearch <usearch>] [--gap-fill {on,off}] [--minpath {on,off}]\n",
      "              [--pathways {metacyc,unipathway}]\n",
      "              [--pathways-database <pathways_database.tsv>] [--xipe {on,off}]\n",
      "              [--annotation-gene-index <3>] [--id-mapping <id_mapping.tsv>]\n",
      "              [--remove-temp-output]\n",
      "              [--log-level {DEBUG,INFO,WARNING,ERROR,CRITICAL}]\n",
      "              [--o-log <sample.log>] [--output-basename <sample_name>]\n",
      "              [--output-format {tsv,biom}] [--output-max-decimals <10>]\n",
      "              [--remove-column-description-output]\n",
      "              [--remove-stratified-output]\n",
      "\n",
      "HUMAnN : HMP Unified Metabolic Analysis Network 3\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "[0] Common settings:\n",
      "  -i <input.fastq>, --input <input.fastq>\n",
      "                        input file of type {fastq,fastq.gz,fasta,fasta.gz,sam,bam,blastm8,genetable,biom} \n",
      "                        [REQUIRED]\n",
      "  -o <output>, --output <output>\n",
      "                        directory to write output files\n",
      "                        [REQUIRED]\n",
      "  --threads <1>         number of threads/processes\n",
      "                        [DEFAULT: 1]\n",
      "  --version             show program's version number and exit\n",
      "\n",
      "[1] Workflow refinement:\n",
      "  -r, --resume          bypass commands if the output files exist\n",
      "  --bypass-nucleotide-index\n",
      "                        bypass the nucleotide index step and run on the indexed ChocoPhlAn database\n",
      "  --bypass-nucleotide-search\n",
      "                        bypass the nucleotide search steps\n",
      "  --bypass-prescreen    bypass the prescreen step and run on the full ChocoPhlAn database\n",
      "  --bypass-translated-search\n",
      "                        bypass the translated search step\n",
      "  --taxonomic-profile <taxonomic_profile.tsv>\n",
      "                        a taxonomic profile (the output file created by metaphlan)\n",
      "                        [DEFAULT: file will be created]\n",
      "  --memory-use {minimum,maximum}\n",
      "                        the amount of memory to use\n",
      "                        [DEFAULT: minimum]\n",
      "  --input-format {fastq,fastq.gz,fasta,fasta.gz,sam,bam,blastm8,genetable,biom}\n",
      "                        the format of the input file\n",
      "                        [DEFAULT: format identified by software]\n",
      "  --search-mode {uniref50,uniref90}\n",
      "                        search for uniref50 or uniref90 gene families\n",
      "                        [DEFAULT: based on translated database selected]\n",
      "  -v, --verbose         additional output is printed\n",
      "\n",
      "[2] Configure tier 1: prescreen:\n",
      "  --metaphlan <metaphlan>\n",
      "                        directory containing the MetaPhlAn software\n",
      "                        [DEFAULT: $PATH]\n",
      "  --metaphlan-options <metaphlan_options>\n",
      "                        options to be provided to the MetaPhlAn software\n",
      "                        [DEFAULT: \"-t rel_ab\"]\n",
      "  --prescreen-threshold <0.01>\n",
      "                        minimum percentage of reads matching a species\n",
      "                        [DEFAULT: 0.01]\n",
      "\n",
      "[3] Configure tier 2: nucleotide search:\n",
      "  --bowtie2 <bowtie2>   directory containing the bowtie2 executable\n",
      "                        [DEFAULT: $PATH]\n",
      "  --bowtie-options <bowtie_options>\n",
      "                        options to be provided to the bowtie software\n",
      "                        [DEFAULT: \"--very-sensitive\"]\n",
      "  --nucleotide-database <nucleotide_database>\n",
      "                        directory containing the nucleotide database\n",
      "                        [DEFAULT: /usr/local/lib/python3.6/dist-packages/humann/data/chocophlan_DEMO]\n",
      "  --nucleotide-identity-threshold <0.0>\n",
      "                        identity threshold for nuclotide alignments\n",
      "                        [DEFAULT: 0.0]\n",
      "  --nucleotide-query-coverage-threshold <90.0>\n",
      "                        query coverage threshold for nucleotide alignments\n",
      "                        [DEFAULT: 90.0]\n",
      "  --nucleotide-subject-coverage-threshold <50.0>\n",
      "                        subject coverage threshold for nucleotide alignments\n",
      "                        [DEFAULT: 50.0]\n",
      "\n",
      "[3] Configure tier 2: translated search:\n",
      "  --diamond <diamond>   directory containing the diamond executable\n",
      "                        [DEFAULT: $PATH]\n",
      "  --diamond-options <diamond_options>\n",
      "                        options to be provided to the diamond software\n",
      "                        [DEFAULT: \"--top 1 --outfmt 6\"]\n",
      "  --evalue <1.0>        the evalue threshold to use with the translated search\n",
      "                        [DEFAULT: 1.0]\n",
      "  --protein-database <protein_database>\n",
      "                        directory containing the protein database\n",
      "                        [DEFAULT: /usr/local/lib/python3.6/dist-packages/humann/data/uniref_DEMO]\n",
      "  --rapsearch <rapsearch>\n",
      "                        directory containing the rapsearch executable\n",
      "                        [DEFAULT: $PATH]\n",
      "  --translated-alignment {usearch,rapsearch,diamond}\n",
      "                        software to use for translated alignment\n",
      "                        [DEFAULT: diamond]\n",
      "  --translated-identity-threshold <Automatically: 50.0 or 80.0, Custom: 0.0-100.0>\n",
      "                        identity threshold for translated alignments\n",
      "                        [DEFAULT: Tuned automatically (based on uniref mode) unless a custom value is specified]\n",
      "  --translated-query-coverage-threshold <90.0>\n",
      "                        query coverage threshold for translated alignments\n",
      "                        [DEFAULT: 90.0]\n",
      "  --translated-subject-coverage-threshold <50.0>\n",
      "                        subject coverage threshold for translated alignments\n",
      "                        [DEFAULT: 50.0]\n",
      "  --usearch <usearch>   directory containing the usearch executable\n",
      "                        [DEFAULT: $PATH]\n",
      "\n",
      "[5] Gene and pathway quantification:\n",
      "  --gap-fill {on,off}   turn on/off the gap fill computation\n",
      "                        [DEFAULT: on]\n",
      "  --minpath {on,off}    turn on/off the minpath computation\n",
      "                        [DEFAULT: on]\n",
      "  --pathways {metacyc,unipathway}\n",
      "                        the database to use for pathway computations\n",
      "                        [DEFAULT: metacyc]\n",
      "  --pathways-database <pathways_database.tsv>\n",
      "                        mapping file (or files, at most two in a comma-delimited list) to use for pathway computations\n",
      "                        [DEFAULT: metacyc database ]\n",
      "  --xipe {on,off}       turn on/off the xipe computation\n",
      "                        [DEFAULT: off]\n",
      "  --annotation-gene-index <3>\n",
      "                        the index of the gene in the sequence annotation\n",
      "                        [DEFAULT: 3]\n",
      "  --id-mapping <id_mapping.tsv>\n",
      "                        id mapping file for alignments\n",
      "                        [DEFAULT: alignment reference used]\n",
      "\n",
      "[6] More output configuration:\n",
      "  --remove-temp-output  remove temp output files\n",
      "                        [DEFAULT: temp files are not removed]\n",
      "  --log-level {DEBUG,INFO,WARNING,ERROR,CRITICAL}\n",
      "                        level of messages to display in log\n",
      "                        [DEFAULT: DEBUG]\n",
      "  --o-log <sample.log>  log file\n",
      "                        [DEFAULT: temp/sample.log]\n",
      "  --output-basename <sample_name>\n",
      "                        the basename for the output files\n",
      "                        [DEFAULT: input file basename]\n",
      "  --output-format {tsv,biom}\n",
      "                        the format of the output files\n",
      "                        [DEFAULT: tsv]\n",
      "  --output-max-decimals <10>\n",
      "                        the number of decimals to output\n",
      "                        [DEFAULT: 10]\n",
      "  --remove-column-description-output\n",
      "                        remove the description in the output column\n",
      "                        [DEFAULT: output column includes description]\n",
      "  --remove-stratified-output\n",
      "                        remove stratification from output\n",
      "                        [DEFAULT: output is stratified]\n"
     ]
    }
   ],
   "source": [
    "# change kernel to Image_humann\n",
    "!humann --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "039d4ea3-4c0c-47b1-817e-529d4eabb95b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Image_humann_latest",
   "language": "python",
   "name": "humann_latest"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
