{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6e5de931-40fc-4c10-86bd-d0f9694ab738",
   "metadata": {},
   "source": [
    "# ğŸ¦  Amplicon Sequencing Data Analysis with QIIME 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54bc1bce-9b47-4165-a09d-fa58f5a640af",
   "metadata": {},
   "source": [
    "## ç’°å¢ƒè¨­å®š"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6dd0a934-cdac-45fa-950b-a90615a7ccce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# å°å¤–é€£ç·šç¶²è·¯è¨­å®š\n",
    "import os\n",
    "tmp=!echo $(hostname)\n",
    "HOSTNAME=tmp[0]\n",
    "os.environ['http_proxy'] = \"socks5:/\"+HOSTNAME+\":12345\" \n",
    "os.environ['https_proxy'] = \"socks5://\"+HOSTNAME+\":12345\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bc3cec47-1589-4a68-9d7f-0da8d9d46a06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# åŸ·è¡Œæª”è·¯å¾‘è¨­å®š\n",
    "import os\n",
    "from pathlib import Path\n",
    "HOME = str(Path.home())\n",
    "Add_Binarry_Path=HOME+'/.local/bin:/usr/localbin'\n",
    "os.environ['PATH']=os.environ['PATH']+':'+Add_Binarry_Path\n",
    "\n",
    "# è‹¥ç”¨Conda è«‹è£œå……ä¸‹åˆ—æŒ‡ä»¤\n",
    "#Add_Binarry_Path=HOME+'/.conda/envs/qiime2-amplicon-2024.10/bin'\n",
    "#os.environ['PATH']=os.environ['PATH']+':'+Add_Binarry_Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1706a61b-f679-4acc-80b4-2fe5c1d9300e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Qiime2 åˆå§‹ç’°å¢ƒè¨­å®š\n",
    "os.environ['MPLCONFIGDIR'] = \"/tmp/mplconfigdir\"\n",
    "os.environ['NUMBA_CACHE_DIR'] = \"/tmp/numbacache\"\n",
    "os.environ['XDG_CONFIG_HOME']=os.environ['HOME']\n",
    "os.environ['CONDA_PREFIX']=os.environ['CONDA_PREFIX'].replace(\"/home/qiime2\", os.environ['HOME'])\n",
    "newpath=os.environ['CONDA_PREFIX']\n",
    "if not os.path.exists(newpath):\n",
    "    os.makedirs(newpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db20665b-41f2-4ad7-8b71-6a03a75613cb",
   "metadata": {},
   "source": [
    "# é–‹å§‹å§ï¼\r\n",
    "\r\n",
    "ç¾åœ¨é€²å…¥æœ‰è¶£çš„éƒ¨åˆ†äº†ã€‚æˆ‘å€‘å…ˆä¾†çœ‹çœ‹æˆ‘å€‘çš„è³‡æ–™ã€‚åœ¨ _data_ è³‡æ–™å¤¾è£¡ï¼Œä½ æœƒæ‰¾åˆ°å…«å€‹ FASTQ æª”æ¡ˆã€ä¸€å€‹æ–‡ä»¶æ¸…å–®ï¼ˆmanifestï¼‰å’Œä¸€å€‹å…ƒæ•¸æ“šæª”æ¡ˆã€‚é¦–å…ˆï¼Œæˆ‘å€‘ä¾†çœ‹çœ‹æ¸…å–®æª”æ¡ˆã€‚é€™æ˜¯ä¸€å€‹åŒ…å«æ‰€æœ‰æ¨£æœ¬åç¨±å’Œæª”æ¡ˆè·¯å¾‘çš„æ–‡ä»¶ï¼Œç¨å¾Œæˆ‘å€‘åœ¨ä½¿ç”¨ QIIME2 æ™‚æœƒéœ€è¦ç”¨åˆ°å®ƒ ğŸ“ã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e1272c1f-b917-4240-b9ad-2a8cc83679e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample-id</th>\n",
       "      <th>absolute-filepath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR1883195</td>\n",
       "      <td>$PWD/data/ERR1883195.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR1883207</td>\n",
       "      <td>$PWD/data/ERR1883207.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR1883212</td>\n",
       "      <td>$PWD/data/ERR1883212.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR1883214</td>\n",
       "      <td>$PWD/data/ERR1883214.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR1883225</td>\n",
       "      <td>$PWD/data/ERR1883225.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERR1883240</td>\n",
       "      <td>$PWD/data/ERR1883240.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ERR1883250</td>\n",
       "      <td>$PWD/data/ERR1883250.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERR1883294</td>\n",
       "      <td>$PWD/data/ERR1883294.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample-id              absolute-filepath\n",
       "0  ERR1883195  $PWD/data/ERR1883195.fastq.gz\n",
       "1  ERR1883207  $PWD/data/ERR1883207.fastq.gz\n",
       "2  ERR1883212  $PWD/data/ERR1883212.fastq.gz\n",
       "3  ERR1883214  $PWD/data/ERR1883214.fastq.gz\n",
       "4  ERR1883225  $PWD/data/ERR1883225.fastq.gz\n",
       "5  ERR1883240  $PWD/data/ERR1883240.fastq.gz\n",
       "6  ERR1883250  $PWD/data/ERR1883250.fastq.gz\n",
       "7  ERR1883294  $PWD/data/ERR1883294.fastq.gz"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "manifest = pd.read_csv('data/manifest.tsv', sep = '\\t')\n",
    "manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "03ab8003-c707-44d9-b4f8-468b34562bd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample-id</th>\n",
       "      <th>collection_timestamp</th>\n",
       "      <th>day_relative_to_fmt</th>\n",
       "      <th>description</th>\n",
       "      <th>disease_state</th>\n",
       "      <th>host_age</th>\n",
       "      <th>host_age_units</th>\n",
       "      <th>host_body_mass_index</th>\n",
       "      <th>host_height</th>\n",
       "      <th>host_height_units</th>\n",
       "      <th>host_subject_id</th>\n",
       "      <th>host_weight</th>\n",
       "      <th>host_weight_units</th>\n",
       "      <th>race</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR1883195</td>\n",
       "      <td>2011-10-24</td>\n",
       "      <td>26</td>\n",
       "      <td>Donor 11</td>\n",
       "      <td>healthy</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>years</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>m</td>\n",
       "      <td>Donor</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>kg</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR1883207</td>\n",
       "      <td>2012-01-12</td>\n",
       "      <td>44</td>\n",
       "      <td>Donor 12</td>\n",
       "      <td>healthy</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>years</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>m</td>\n",
       "      <td>Donor</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>kg</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR1883212</td>\n",
       "      <td>2012-10-10</td>\n",
       "      <td>135</td>\n",
       "      <td>Donor 14</td>\n",
       "      <td>healthy</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>years</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>m</td>\n",
       "      <td>Donor</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>kg</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR1883214</td>\n",
       "      <td>2011-07-26</td>\n",
       "      <td>0</td>\n",
       "      <td>Day 0 CD1</td>\n",
       "      <td>Pre-FMT</td>\n",
       "      <td>39</td>\n",
       "      <td>years</td>\n",
       "      <td>29.3</td>\n",
       "      <td>165.1</td>\n",
       "      <td>m</td>\n",
       "      <td>CD1</td>\n",
       "      <td>80.1</td>\n",
       "      <td>kg</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR1883225</td>\n",
       "      <td>2011-07-26</td>\n",
       "      <td>54</td>\n",
       "      <td>Donor CD1</td>\n",
       "      <td>healthy</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>years</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>m</td>\n",
       "      <td>Donor</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>kg</td>\n",
       "      <td>Restricted access</td>\n",
       "      <td>Restricted access</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERR1883240</td>\n",
       "      <td>2012-02-14</td>\n",
       "      <td>pre-FMT</td>\n",
       "      <td>CD9 pre-FMT</td>\n",
       "      <td>Pre-FMT</td>\n",
       "      <td>47</td>\n",
       "      <td>years</td>\n",
       "      <td>35.5</td>\n",
       "      <td>1.55</td>\n",
       "      <td>m</td>\n",
       "      <td>CD9</td>\n",
       "      <td>85.1</td>\n",
       "      <td>kg</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ERR1883250</td>\n",
       "      <td>2011-12-23</td>\n",
       "      <td>pre-FMT</td>\n",
       "      <td>CD13 pre-FMT</td>\n",
       "      <td>Pre-FMT</td>\n",
       "      <td>53</td>\n",
       "      <td>years</td>\n",
       "      <td>34.4</td>\n",
       "      <td>1.56</td>\n",
       "      <td>m</td>\n",
       "      <td>CD13</td>\n",
       "      <td>83.9</td>\n",
       "      <td>kg</td>\n",
       "      <td>white</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERR1883294</td>\n",
       "      <td>2011-09-29</td>\n",
       "      <td>0</td>\n",
       "      <td>Day 0 CD3</td>\n",
       "      <td>Pre-FMT</td>\n",
       "      <td>61</td>\n",
       "      <td>years</td>\n",
       "      <td>32.5</td>\n",
       "      <td>1.727</td>\n",
       "      <td>m</td>\n",
       "      <td>CD3</td>\n",
       "      <td>97.3</td>\n",
       "      <td>kg</td>\n",
       "      <td>white</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample-id collection_timestamp day_relative_to_fmt   description  \\\n",
       "0  ERR1883195           2011-10-24                  26      Donor 11   \n",
       "1  ERR1883207           2012-01-12                  44      Donor 12   \n",
       "2  ERR1883212           2012-10-10                 135      Donor 14   \n",
       "3  ERR1883214           2011-07-26                   0     Day 0 CD1   \n",
       "4  ERR1883225           2011-07-26                  54     Donor CD1   \n",
       "5  ERR1883240           2012-02-14             pre-FMT   CD9 pre-FMT   \n",
       "6  ERR1883250           2011-12-23             pre-FMT  CD13 pre-FMT   \n",
       "7  ERR1883294           2011-09-29                   0     Day 0 CD3   \n",
       "\n",
       "  disease_state           host_age host_age_units host_body_mass_index  \\\n",
       "0       healthy  Restricted access          years    Restricted access   \n",
       "1       healthy  Restricted access          years    Restricted access   \n",
       "2       healthy  Restricted access          years    Restricted access   \n",
       "3       Pre-FMT                 39          years                 29.3   \n",
       "4       healthy  Restricted access          years    Restricted access   \n",
       "5       Pre-FMT                 47          years                 35.5   \n",
       "6       Pre-FMT                 53          years                 34.4   \n",
       "7       Pre-FMT                 61          years                 32.5   \n",
       "\n",
       "         host_height host_height_units host_subject_id        host_weight  \\\n",
       "0  Restricted access                 m           Donor  Restricted access   \n",
       "1  Restricted access                 m           Donor  Restricted access   \n",
       "2  Restricted access                 m           Donor  Restricted access   \n",
       "3              165.1                 m             CD1               80.1   \n",
       "4  Restricted access                 m           Donor  Restricted access   \n",
       "5               1.55                 m             CD9               85.1   \n",
       "6               1.56                 m            CD13               83.9   \n",
       "7              1.727                 m             CD3               97.3   \n",
       "\n",
       "  host_weight_units               race                sex  \n",
       "0                kg  Restricted access  Restricted access  \n",
       "1                kg  Restricted access  Restricted access  \n",
       "2                kg  Restricted access  Restricted access  \n",
       "3                kg              white             female  \n",
       "4                kg  Restricted access  Restricted access  \n",
       "5                kg              white             female  \n",
       "6                kg              white             female  \n",
       "7                kg              white               male  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv('data/metadata.tsv', sep = '\\t')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2590c3c7-bb43-49b5-b74a-921ebb0eb012",
   "metadata": {},
   "source": [
    "çœ‹èµ·ä¾†ä¸éŒ¯ï¼Œæ‰€æœ‰å…«å€‹ FASTQ æª”æ¡ˆéƒ½å·²ç¢ºèªç„¡èª¤ï¼Œå››å€‹æ˜¯å¥åº·æ¨£æœ¬ï¼Œå››å€‹æ˜¯åè¦†æ€§CDIçš„æ¨£æœ¬ã€‚æˆ‘å€‘å¯ä»¥ä½¿ç”¨æ¸…å–®æª”å°‡æˆ‘å€‘çš„æª”æ¡ˆåŒ¯å…¥ QIIME2ã€‚\r\n",
    "\r\n",
    "## QIIME2 æµç¨‹\r\n",
    "\r\n",
    "è®“æˆ‘å€‘å›é¡§ä¸€ä¸‹ QIIME2 æµç¨‹å°‡æœƒåšä»€éº¼ï¼š\n",
    "![our workflow](https://github.com/Gibbons-Lab/isb_course_2023/raw/main/docs/16S/assets/steps.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc3b71fc-e9f2-459a-8b0f-3ade32d677f4",
   "metadata": {},
   "source": [
    "è¦åœ¨ QIIME2 ä¸­ä½¿ç”¨å®šåºè³‡æ–™ï¼Œæˆ‘å€‘é¦–å…ˆéœ€è¦å°‡åŒ…å«æˆ‘å€‘è³‡æ–™çš„ FASTQ æª”æ¡ˆè½‰æ›æˆ QIIME å·¥ä»¶ã€‚ä½¿ç”¨æˆ‘å€‘å‰›æª¢æŸ¥éçš„æ¸…å–®ï¼Œè®“æˆ‘å€‘ä¾†åŸ·è¡Œç¬¬ä¸€å€‹æŒ‡ä»¤ï¼š\r\n",
    "\r\n",
    "-- æé†’ä¸€ä¸‹ï¼Œåœ¨æŒ‡ä»¤å‰åŠ ä¸Š ```!``` è¡¨ç¤ºé€™æ˜¯ä¸€å€‹ bash æŒ‡ä»¤ï¼Œè€Œä¸æ˜¯ pythonã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1c506ab9-b062-4c5e-8539-a3c2e2a366ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mImported data/manifest.tsv as SingleEndFastqManifestPhred33V2 to output/sequences.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "# fastqæª”æ¡ˆæ ¼å¼è½‰æ›æˆqza\n",
    "!mkdir -p output\n",
    "!qiime tools import \\\n",
    "  --type 'SampleData[SequencesWithQuality]' \\\n",
    "  --input-path data/manifest.tsv \\\n",
    "  --output-path output/sequences.qza \\\n",
    "  --input-format SingleEndFastqManifestPhred33V2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b53f3bc-eda4-4add-84f3-818f151dcab4",
   "metadata": {},
   "source": [
    "## ç¢ºèªqzaæª”æ¡ˆå…§å®¹ç‰©"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e9685b8c-8c91-4866-af2c-c9be4f8bfb7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mUUID\u001b[0m:        d6b8db49-5596-4976-a33b-03b48d6702ad\n",
      "\u001b[32mType\u001b[0m:        SampleData[SequencesWithQuality]\n",
      "\u001b[32mData format\u001b[0m: SingleLanePerSampleSingleEndFastqDirFmt\n"
     ]
    }
   ],
   "source": [
    "# ç¢ºèªqzaæª”æ¡ˆå…§å®¹ç‰©\n",
    "!qiime tools peek output/sequences.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b06e841-35d8-43ea-83ef-3b644508778d",
   "metadata": {},
   "source": [
    "## è¦–è¦ºåŒ–æˆ‘å€‘çš„æ•¸æ“š ğŸ”\r\n",
    "\r\n",
    "åœ¨æˆ‘å€‘ç¹¼çºŒä¹‹å‰ï¼Œè®“æˆ‘å€‘ä½¿ç”¨ QIIME2 ä¾†è¦–è¦ºåŒ–æˆ‘å€‘çš„æ¸¬åºæ•¸æ“šã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "467f3789-f882-49f0-a3ab-256f8fc4c207",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: output/demux.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!cp output/sequences.qza output/demux.qza\n",
    "!qiime demux summarize \\\n",
    "--i-data output/demux.qza \\\n",
    "--o-visualization output/demux.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a44ff0c-bd34-4f4f-853e-5ed552f915c8",
   "metadata": {},
   "source": [
    ".qzv æª”æ¡ˆåƒæˆ‘å€‘å‰›å‰›ç”¢ç”Ÿçš„é€™å€‹ï¼Œæ˜¯ç”¨ä¾†è¦–è¦ºåŒ–çš„æª”æ¡ˆã€‚ä½ å¯ä»¥ä¸‹è¼‰é€™å€‹æª”æ¡ˆï¼Œç„¶å¾Œä½¿ç”¨ http://view.qiime2.org æ‰“é–‹ä¾†æŸ¥çœ‹åœ–è¡¨ã€‚è¦ä¸‹è¼‰æª”æ¡ˆï¼Œé»æ“Šå·¦å´çš„è³‡æ–™å¤¾ç¬¦è™Ÿï¼Œæ‰“é–‹ `output` è³‡æ–™å¤¾ï¼Œç„¶å¾Œåœ¨ `output/demux.qzv` æª”æ¡ˆæ—é‚Šçš„é»é¸å–®ä¸­é¸æ“‡ä¸‹è¼‰ã€‚\r\n",
    "\r\n",
    "---\r\n",
    "\r\n",
    "## è³ªé‡éæ¿¾\r\n",
    "\r\n",
    "åœ¨ä½¿ç”¨æˆ‘å€‘çš„å®šåºæ•¸æ“šä¹‹å‰ï¼Œæˆ‘å€‘éœ€è¦å°å…¶é€²è¡Œâ€œå»å™ªâ€è™•ç†ã€‚ç‚ºæ­¤ï¼Œæˆ‘å€‘å°‡ä½¿ç”¨ä¸€å€‹åç‚º DADA2 çš„æ’ä»¶ã€‚é€™å€‹éç¨‹åŒ…å«ä¸‰å€‹æ­¥é©Ÿï¼š\r\n",
    "\r\n",
    "1. éæ¿¾ä¸¦ä¿®å‰ªè®€å–åºåˆ—\r\n",
    "2. æ‰¾å‡ºæ¨£æœ¬ä¸­æœ€æœ‰å¯èƒ½çš„å”¯ä¸€åºåˆ—é›† (ASVs)\r\n",
    "3. ç§»é™¤åµŒåˆé«”\r\n",
    "4. è¨ˆç°¡å ±ä¸­è¨è«–æ­£åœ¨ç™¼ç”Ÿçš„äº‹æƒ…ã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e53c2de6-1410-4f93-bcb8-9ebc8865424d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: output/table.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: output/rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[DADA2Stats] to: output/stats.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime dada2 denoise-single \\\n",
    "  --i-demultiplexed-seqs output/demux.qza \\\n",
    "  --p-trim-left 0 \\\n",
    "  --p-trunc-len 150 \\\n",
    "  --p-n-threads 8 \\\n",
    "  --o-representative-sequences output/rep-seqs.qza \\\n",
    "  --o-table output/table.qza \\\n",
    "  --o-denoising-stats output/stats.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee42001f-423c-4978-87dc-d5faecbf0224",
   "metadata": {},
   "source": [
    "è®“æˆ‘å€‘ä¾†æª¢æŸ¥ä¸€ä¸‹çµæœå¦‚ä½•ã€‚åˆ¤æ–·æ‰€è­˜åˆ¥çš„ ASVs æ˜¯å¦èƒ½ä»£è¡¨æ¨£æœ¬çš„ä¸€å€‹å¥½æ–¹æ³•æ˜¯æª¢æŸ¥åœ¨æ•´å€‹åˆ†ææµç¨‹ä¸­ä¿ç•™äº†å¤šå°‘è®€æ•¸ã€‚ä»¥ä¸‹æ˜¯å¸¸è¦‹å•é¡ŒåŠå…¶è§£æ±ºæ–¹æ¡ˆï¼š\r\n",
    "\r\n",
    "**åœ¨åˆä½µéç¨‹ä¸­å¤§é‡è®€æ•¸ä¸Ÿå¤±ï¼ˆåƒ…é™é›™ç«¯æ¸¬åºï¼‰**\r\n",
    "\r\n",
    "![è®€æ•¸é‡ç–Š](https://gibbons-lab.github.io/isb_course_2023/16S/assets/read_overlap.png)\r\n",
    "\r\n",
    "DADA2 åœ¨åˆä½µ ASVs æ™‚ï¼Œé è¨­ä½¿ç”¨å‰å‘è®€æ•¸å’Œåå‘è®€æ•¸ä¹‹é–“æœ‰ 12 å€‹é¹¼åŸºçš„é‡ç–Šã€‚å› æ­¤ï¼Œè®€æ•¸åœ¨ä¿®å‰ªå¾Œå¿…é ˆå…è¨±è¶³å¤ çš„é‡ç–Šã€‚å¦‚æœæ“´å¢å€åŸŸé•·åº¦ç‚º 450bpï¼Œè€Œä½ æœ‰ 2x250bp çš„è®€æ•¸ä¸¦ä¿®å‰ªäº†æ¯å€‹è®€æ•¸çš„æœ€å¾Œ 30 å€‹é¹¼åŸºï¼Œå°‡è®€æ•¸é•·åº¦ç¸®çŸ­ç‚º 220bpï¼Œé‚£éº¼ç¸½å…±è¦†è“‹çš„åºåˆ—é•·åº¦ç‚º 2x220 = 440bpï¼Œé€™æ¯” 450bp çŸ­ï¼Œå› æ­¤ä¸æœƒæœ‰é‡ç–Šã€‚è¦è§£æ±ºæ­¤å•é¡Œï¼Œå¯ä»¥æ¸›å°‘è®€æ•¸çš„ä¿®å‰ªé‡ï¼Œæˆ–è€…èª¿æ•´ `--p-min-overlap` åƒæ•¸ç‚ºè¼ƒä½çš„å€¼ï¼ˆä½†ä¸èƒ½å¤ªä½ï¼‰ã€‚\r\n",
    "\r\n",
    "<br>\r\n",
    "\r\n",
    "**å¤§éƒ¨åˆ†è®€æ•¸å› ç‚ºåµŒåˆé«”è€Œä¸Ÿå¤±**\r\n",
    "\r\n",
    "![è®€æ•¸é‡ç–Š](https://gibbons-lab.github.io/isb_course_2023/16S/assets/chimera.png)\r\n",
    "\r\n",
    "é€™é€šå¸¸æ˜¯å¯¦é©—å•é¡Œï¼Œå› ç‚ºåµŒåˆé«”æ˜¯åœ¨æ“´å¢éç¨‹ä¸­ç”¢ç”Ÿçš„ã€‚å¦‚æœå¯ä»¥èª¿æ•´ä½ çš„ PCRï¼Œå˜—è©¦æ¸›å°‘å¾ªç’°æ¬¡æ•¸ã€‚åµŒåˆé«”ä¹Ÿå¯èƒ½ç”±æ–¼éŒ¯èª¤çš„åˆä½µéç¨‹ç”¢ç”Ÿã€‚å¦‚æœæœ€å°é‡ç–Šéå°ï¼ŒASVs å¯èƒ½æœƒè¢«éš¨æ©Ÿåˆä½µã€‚å¯èƒ½çš„è§£æ±ºæ–¹æ³•æ˜¯å¢åŠ  `--p-min-overlap` åƒæ•¸ï¼Œæˆ–åƒ…å°å‰å‘è®€æ•¸é€²è¡Œåˆ†æï¼ˆæ ¹æ“šæˆ‘å€‘çš„ç¶“é©—è§€å¯Ÿï¼ŒåµŒåˆé«”æ›´å¯èƒ½åœ¨åˆä½µçš„è®€æ•¸ä¸­ç”¢ç”Ÿï¼‰ã€‚*ç„¶è€Œï¼Œä¸Ÿå¤± 5-25% çš„è®€æ•¸å› ç‚ºåµŒåˆé«”æ˜¯æ­£å¸¸çš„ï¼Œä¸éœ€è¦åšä»»ä½•èª¿æ•´ã€‚*\r\n",
    "\r\n",
    "æˆ‘å€‘çš„å»å™ªçµ±è¨ˆæ•¸æ“šåŒ…å«åœ¨ä¸€å€‹ç”¢ç‰©ä¸­ã€‚è¦å°‡å…¶è½‰æ›ç‚ºå¯è¦–åŒ–çµæœï¼Œæˆ‘å€‘å¯ä»¥ä½¿ç”¨ `qiime metadata tabulate`ã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b14e884c-88d0-40b7-a8d5-ea4ffdbc5b40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: output/rep-seqs.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: output/table.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: output/stats.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-table tabulate-seqs \\\n",
    "  --i-data output/rep-seqs.qza \\\n",
    "  --o-visualization output/rep-seqs.qzv\n",
    "\n",
    "!qiime feature-table summarize \\\n",
    "  --i-table output/table.qza \\\n",
    "  --m-sample-metadata-file data/metadata.tsv \\\n",
    "  --o-visualization output/table.qzv\n",
    "\n",
    "!qiime metadata tabulate \\\n",
    "    --m-input-file output/stats.qza \\\n",
    "    --o-visualization output/stats.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c227ccc5-1deb-455a-acb2-db6afb4bffd4",
   "metadata": {},
   "source": [
    "åƒä¹‹å‰ä¸€æ¨£ï¼Œæˆ‘å€‘å¯ä»¥ä¸‹è¼‰ .qzv æª”æ¡ˆï¼Œä¸¦ä½¿ç”¨ [QIIME2 Viewer]('https://view.qiime2.org/') ä¾†è¦–è¦ºåŒ–çµæœã€‚\r\n",
    "\r\n",
    "äº†è§£é€™äº›è¼¸å‡ºå…§å®¹å°æˆ‘å€‘ä¾†èªªå¾ˆé‡è¦ã€‚ä¾‹å¦‚ï¼Œæ•¸æ“šä¸­æœ‰å¤šå°‘ç™¾åˆ†æ¯”çš„è®€åºé€šéäº†éæ¿¾æ­¥é©Ÿï¼Ÿæœ‰å¤šå°‘ç™¾åˆ†æ¯”çš„è®€åºæ˜¯éåµŒåˆçš„ï¼Ÿé€™äº›æŒ‡æ¨™åœ¨æ¨£æœ¬ä¹‹é–“çš„å·®ç•°å¯èƒ½æœƒå½±éŸ¿å¤šæ¨£æ€§æŒ‡æ¨™ã€‚"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "240b8f6d-8865-4041-8a31-0ebbb366ef84",
   "metadata": {},
   "source": [
    "## å¤šæ¨£æ€§èˆ‡ç³»çµ±ç™¼è‚²  Diversity and Phylogenetics\r\n",
    "åœ¨ç ”ç©¶å¾®ç”Ÿç‰©ç”Ÿæ…‹å­¸æ™‚ï¼Œä¸€å€‹é‡è¦çš„æŒ‡æ¨™æ˜¯__å¤šæ¨£æ€§__ã€‚å¤šæ¨£æ€§ä¸»è¦åˆ†ç‚ºå…©ç¨®ï¼šâºï¼ˆalphaï¼‰å¤šæ¨£æ€§å’ŒÎ²ï¼ˆbetaï¼‰å¤šæ¨£æ€§ã€‚\r\n",
    "\r\n",
    "Alpha å¤šæ¨£æ€§ç›¸å°ç°¡å–®â€”â€”æŒ‡çš„æ˜¯å–®ä¸€æ¨£æœ¬çš„å¤šæ¨£æ€§ã€‚ä½ å¯ä»¥è€ƒæ…®åƒæ˜¯ç‰©ç¨®è±å¯Œåº¦å’Œå‡å‹»åº¦ç­‰æŒ‡æ¨™ã€‚\r\n",
    "\r\n",
    "![alpha å¤šæ¨£æ€§](https://gibbons-lab.github.io/isb_course_2023/16S/assets/alpha_diversity.png)\r\n",
    "\r\n",
    "Beta å¤šæ¨£æ€§å‰‡æ˜¯çœ‹å…©å€‹æ¨£æœ¬ä¹‹é–“çš„å·®ç•°â€”â€”æœ‰å“ªäº›åˆ†é¡å–®å…ƒæ˜¯å…±äº«çš„ï¼Œä»¥åŠå®ƒå€‘çš„è±\n",
    "æœ‰ä»€éº¼ä¸åŒã€‚\r\n",
    "\r\n",
    "![beta å¤šæ¨£æ€§](https://gibbons-lab.github.io/isb_course_2023/16S/assets/beta_diversit\n",
    "##ng)\r\n",
    "\r\n",
    "\r\n",
    " Starting our Tree### é–‹å§‹å»ºç«‹æˆ‘å€‘çš„æ¨¹\r\n",
    "æ¥ä¸‹ä¾†ï¼Œæˆ‘å€‘å°‡ä½¿ç”¨ä»¥ä¸‹æŒ‡ä»¤ä¾†å»ºç«‹æˆ‘å€‘åºåˆ—çš„ç³»çµ±ç™¼è‚²æ¨¹ã€‚é€™æ¬¡ï¼Œæˆ‘å€‘å°‡èª¿ç”¨ QIIME2 ä¸­çš„ _phylogeny_ æ’ä»¶ã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2230441c-2732-4e33-9829-ffb5071b39e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[AlignedSequence] to: output/aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: output/masked-aligned-rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Unrooted] to: output/unrooted-tree.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Rooted] to: output/rooted-tree.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences output/rep-seqs.qza \\\n",
    "  --o-alignment output/aligned-rep-seqs.qza \\\n",
    "  --o-masked-alignment output/masked-aligned-rep-seqs.qza \\\n",
    "  --o-tree output/unrooted-tree.qza \\\n",
    "  --o-rooted-tree output/rooted-tree.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5542968-75d3-485a-bfbd-1df0ead678b8",
   "metadata": {},
   "source": [
    "## è¨ˆç®—å¤šæ¨£æ€§\r\n",
    "ä½¿ç”¨å¤šæ¨£æ€§æ’ä»¶ï¼Œæˆ‘å€‘å¯ä»¥åˆ©ç”¨è¡¨æ ¼å’Œæ¨¹ç‹€åœ–ä¾†è¨ˆç®—å¹¾å€‹å¤šæ¨£æ€§æŒ‡æ¨™ã€‚ç‚ºäº†è€ƒæ…®æ¨£æœ¬æ·±åº¦çš„è®Šç•°ï¼Œæˆ‘å€‘æœƒåœ¨ QIIME2 ä¸­è¨­ç½®ä¸€å€‹æˆªæ­¢é»ï¼Œä¸¦åœ¨è©²é»å°æ‰€æœ‰æ¨£æœ¬é€²è¡Œç¨€é‡‹ã€‚ç”±æ–¼é€™æ˜¯éš¨æ©Ÿé¸å–åºåˆ—ï¼Œå› æ­¤çµæœå¯èƒ½æœƒæœ‰æ‰€ä¸åŒã€‚æˆ‘å€‘é‚„æœƒè¼¸å…¥å…ƒæ•¸æ“šæ–‡ä»¶ï¼Œä»¥ä¾¿è¿½è¹¤æ¯å€‹æ¨£æœ¬å±¬æ–¼å“ªå€‹çµ„åˆ¥ã€‚è¼¸å…¥å…ƒæ•¸æ“šæ–‡ä»¶ï¼Œä»¥ä¾¿è¿½è¹¤æ¯å€‹æ¨£æœ¬å±¬æ–¼å“ªå€‹çµ„åˆ¥ã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "777d4036-42e4-4176-8d81-9c7ae811f1f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: diversity/rarefied_table.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: diversity/faith_pd_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: diversity/observed_features_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: diversity/shannon_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: diversity/evenness_vector.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: diversity/unweighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: diversity/weighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: diversity/jaccard_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: diversity/bray_curtis_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: diversity/unweighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: diversity/weighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: diversity/jaccard_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: diversity/bray_curtis_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved Visualization to: diversity/unweighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: diversity/weighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: diversity/jaccard_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: diversity/bray_curtis_emperor.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity core-metrics-phylogenetic \\\n",
    "    --i-table output/table.qza \\\n",
    "    --i-phylogeny output/rooted-tree.qza \\\n",
    "    --p-sampling-depth 8000 \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --output-dir diversity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df764e76-1183-4b75-b08c-2d17f96558d8",
   "metadata": {},
   "source": [
    "## Alpha Diversity\n",
    "\n",
    "We get a bunch of outputs from the previous command - measures of both alpha and beta diversity. To start, let's use the Shannon vector in the output directory to create a visualization of alpha diversity across samples. Generally, healthy, long-living individuals have balanced diverse microbiomes. However, this isn't necessarily a direct indicator of health or disease. Let's see how it looks in our samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23bb804a-786f-4615-a7a9-d3488dd8b94d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: diversity/alpha_groups-shannon_vector.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "    --i-alpha-diversity diversity/shannon_vector.qza \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --o-visualization diversity/alpha_groups-shannon_vector.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0d9c9143-8710-417f-8531-bf527b929678",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: diversity/alpha_groups-faith_pd_vector.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "    --i-alpha-diversity diversity/faith_pd_vector.qza \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --o-visualization diversity/alpha_groups-faith_pd_vector.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8117d40f-adc2-440e-812c-874bc818a963",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: diversity/alpha_groups-evenness_vector.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "    --i-alpha-diversity diversity/evenness_vector.qza \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --o-visualization diversity/alpha_groups-evenness_vector.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "081b5b75-88ca-4fbc-85f9-3153bf1d817f",
   "metadata": {},
   "source": [
    "åƒä¹‹å‰ä¸€æ¨£ï¼Œæˆ‘å€‘å¯ä»¥ä¸‹è¼‰è¦–è¦ºåŒ–çµæœä¸¦ç”¨ QIIME2 æŸ¥çœ‹å™¨æ‰“é–‹å®ƒã€‚\r\n",
    "\r\n",
    "## Beta å¤šæ¨£æ€§\r\n",
    "\r\n",
    "è®“æˆ‘å€‘ä¾†è¦–è¦ºåŒ– Beta å¤šæ¨£æ€§ï¼Œçœ‹çœ‹å®ƒå€‘æ˜¯å¦‚ä½•å€åˆ†çš„ã€‚é€™æ¬¡æˆ‘å€‘å°‡ä½¿ç”¨åŠ æ¬Š UniFracã€‚æˆ‘å€‘éœ€è¦ä¸‹è¼‰é€™å€‹æª”æ¡ˆ â¬…ï¸\r\n",
    "\r\n",
    "<br>\r\n",
    "\r\n",
    "æˆ‘å€‘å¯ä»¥ä½¿ç”¨ PERMANOVA æª¢æŸ¥æ¨£æœ¬ä¹‹é–“æ˜¯å¦æœ‰ã€Œé¡¯è‘—ã€çš„å€åˆ†ã€‚æˆ‘å€‘å¯ä»¥é€é QIIME2 çš„å¤šæ¨£æ€§æ’ä»¶ä¾†å®Œæˆé€™é …æ“ä½œã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "47e6dedd-7f1a-42d7-b025-1a5abdb3305e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: diversity/permanova.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity adonis \\\n",
    "    --i-distance-matrix diversity/weighted_unifrac_distance_matrix.qza \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --p-formula \"disease_state\" \\\n",
    "    --p-n-jobs 2 \\\n",
    "    --o-visualization diversity/permanova.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6eef3d8d-1748-4f8a-ad19-b9d9caea82ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: diversity/beta_groups-weighted_unifrac_distance_matrix.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "    --i-distance-matrix diversity/weighted_unifrac_distance_matrix.qza \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --m-metadata-column disease_state \\\n",
    "    --o-visualization diversity/beta_groups-weighted_unifrac_distance_matrix.qzv \\\n",
    "    --p-pairwise\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "98666d32-3e0b-4537-8df6-6ab194f8a42f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: diversity/beta_groups-weighted_unifrac_distance_matrix.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "    --i-distance-matrix diversity/weighted_unifrac_distance_matrix.qza \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --m-metadata-column disease_state \\\n",
    "    --o-visualization diversity/beta_groups-weighted_unifrac_distance_matrix.qzv \\\n",
    "    --p-pairwise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65f206ee-14b7-468f-904e-8a54e00af1fa",
   "metadata": {},
   "source": [
    "## ç³»çµ±åˆ†é¡\r\n",
    "\r\n",
    "æˆ‘å€‘å¯ä»¥å¾å¤šæ¨£æ€§æŒ‡æ¨™ã€Î±å¤šæ¨£æ€§å’ŒÎ²å¤šæ¨£æ€§ä¸­å­¸åˆ°å¾ˆå¤šæ±è¥¿ã€‚ä½†è¦çœŸæ­£æ·±å…¥äº†è§£æ•¸æ“šï¼Œæˆ‘å€‘éœ€è¦çŸ¥é“æ¯å€‹æ¨£æœ¬ä¸­æœ‰å“ªäº›å¾®ç”Ÿç‰© ğŸ¦ ã€‚ç‚ºæ­¤ï¼Œæˆ‘å€‘å°‡ä½¿ç”¨è²è‘‰æ–¯åˆ†é¡å™¨åœ¨ QIIME2 ä¸­å°è®€æ•¸é€²è¡Œåˆ†é¡ã€‚å¯ä»¥åœ¨ https://docs.qiime2.org/2425.7/data-resources æ‰¾åˆ°å¹¾ç¨®é€™æ¨£çš„åˆ†é¡å™¨ã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dd7e4cb9-18c2-4637-a68d-8e82fea6bb0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "!curl -sL \\\n",
    "  \"https://data.qiime2.org/classifiers/sklearn-1.4.2/greengenes/gg-13-8-99-515-806-nb-classifier.qza\" > \\\n",
    "  \"output/gg-13-8-99-515-806-nb-classifier.qza\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7f40d9ee-9352-42c6-abaf-a500fa65c029",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[Taxonomy] to: output/taxonomy.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-classifier classify-sklearn \\\n",
    "    --i-reads output/rep-seqs.qza \\\n",
    "    --i-classifier output/gg-13-8-99-515-806-nb-classifier.qza \\\n",
    "    --p-n-jobs 2 \\\n",
    "    --o-classification output/taxonomy.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "885f7cad-62c2-4eb5-a4c9-0f47f9c6448c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: output/taxonomy.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime metadata tabulate \\\n",
    "  --m-input-file output/taxonomy.qza \\\n",
    "  --o-visualization output/taxonomy.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0801cf0d-7dcd-4594-a9c4-cb8dbf053132",
   "metadata": {},
   "source": [
    "ç¾åœ¨æˆ‘å€‘å·²ç¶“å°‡è®€å–è³‡æ–™åˆ†é¡ï¼Œæˆ‘å€‘å¯ä»¥è¦–è¦ºåŒ–æˆ‘å€‘æ¨£æœ¬çš„åˆ†é¡å­¸åˆ†ä½ˆã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cd5c2eec-b01f-42fb-a6e1-344264440b56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: output/taxa_barplot.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime taxa barplot \\\n",
    "    --i-table output/table.qza \\\n",
    "    --i-taxonomy output/taxonomy.qza \\\n",
    "    --m-metadata-file data/metadata.tsv \\\n",
    "    --o-visualization output/taxa_barplot.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55051074-7c83-4db8-981d-b87076c60843",
   "metadata": {},
   "source": [
    "ç¾åœ¨ï¼Œæˆ‘å€‘å¯ä»¥ä½¿ç”¨åŒ…å«æˆ‘å€‘è®€å–è³‡æ–™çš„ ```table.qza``` å’ŒåŒ…å«è®€å–è³‡æ–™åˆ†é¡è³‡è¨Šçš„ ```taxa.qza```ï¼Œå°‡è³‡æ–™æ•´åˆåˆ°å±¬ (genus) çš„å±¤ç´šã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1129ec67-f5e9-49d8-af0a-124e824cde0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: output/genus.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime taxa collapse \\\n",
    "    --i-table output/table.qza \\\n",
    "    --i-taxonomy output/taxonomy.qza \\\n",
    "    --p-level 6 \\\n",
    "    --o-collapsed-table output/genus.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f21d115-21b2-41fe-a4e6-11e8d9f69492",
   "metadata": {},
   "source": [
    "æˆ‘å€‘æœƒå°‡é€™å€‹åŒ¯å‡ºç‚º .tsv æ ¼å¼ï¼Œé€™æ¨£å°æ–¼èª²ç¨‹çš„ä¸‹ä¸€éƒ¨åˆ†æœƒæ›´æ–¹ä¾¿ä½¿ç”¨ã€‚"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c4927dc3-bd5f-4277-87d7-0d9fa333463a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mExported output/genus.qza as BIOMV210DirFmt to directory exported\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime tools export \\\n",
    "    --input-path output/genus.qza \\\n",
    "    --output-path exported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "366d451e-2c77-4df2-a043-3464a49da559",
   "metadata": {},
   "outputs": [],
   "source": [
    "!biom convert -i exported/feature-table.biom -o exported/genus.tsv --to-tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "90e72e75-33b0-4b2d-84c1-cf8515445292",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ERR1883195</th>\n",
       "      <th>ERR1883207</th>\n",
       "      <th>ERR1883212</th>\n",
       "      <th>ERR1883214</th>\n",
       "      <th>ERR1883225</th>\n",
       "      <th>ERR1883240</th>\n",
       "      <th>ERR1883250</th>\n",
       "      <th>ERR1883294</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#OTU ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae;g__Escherichia</th>\n",
       "      <td>23.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>51036.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1779.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>9014.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Verrucomicrobiaceae;g__Akkermansia</th>\n",
       "      <td>6.0</td>\n",
       "      <td>598.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54728.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35554.0</td>\n",
       "      <td>11467.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides</th>\n",
       "      <td>24784.0</td>\n",
       "      <td>23969.0</td>\n",
       "      <td>11679.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>5209.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae;__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1553.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15863.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>9230.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Veillonellaceae;g__Dialister</th>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2650.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15826.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1199.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Dehalobacteriaceae;g__Dehalobacterium</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__Streptophyta;f__;g__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__[Tissierellaceae];g__WAL_1855D</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__TM7;c__TM7-3;o__Blgi18;f__;g__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__[Mogibacteriaceae];g__Mogibacterium</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ERR1883195  ERR1883207  \\\n",
       "#OTU ID                                                                      \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...        23.0        37.0   \n",
       "k__Bacteria;p__Verrucomicrobia;c__Verrucomicrob...         6.0       598.0   \n",
       "k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__...     24784.0     23969.0   \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...         0.0         0.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         5.0         3.0   \n",
       "...                                                        ...         ...   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         2.0   \n",
       "k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__...         0.0         0.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0   \n",
       "k__Bacteria;p__TM7;c__TM7-3;o__Blgi18;f__;g__              0.0         0.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0   \n",
       "\n",
       "                                                    ERR1883212  ERR1883214  \\\n",
       "#OTU ID                                                                      \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...        14.0     51036.0   \n",
       "k__Bacteria;p__Verrucomicrobia;c__Verrucomicrob...         0.0     54728.0   \n",
       "k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__...     11679.0        59.0   \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...        27.0      1553.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...        17.0      2650.0   \n",
       "...                                                        ...         ...   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0   \n",
       "k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__...         2.0         0.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         2.0         0.0   \n",
       "k__Bacteria;p__TM7;c__TM7-3;o__Blgi18;f__;g__              2.0         0.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0   \n",
       "\n",
       "                                                    ERR1883225  ERR1883240  \\\n",
       "#OTU ID                                                                      \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...         0.0      1779.0   \n",
       "k__Bacteria;p__Verrucomicrobia;c__Verrucomicrob...         0.0     35554.0   \n",
       "k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__...      5209.0        39.0   \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...         0.0     15863.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0     15826.0   \n",
       "...                                                        ...         ...   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0   \n",
       "k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__...         0.0         0.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0   \n",
       "k__Bacteria;p__TM7;c__TM7-3;o__Blgi18;f__;g__              0.0         0.0   \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0   \n",
       "\n",
       "                                                    ERR1883250  ERR1883294  \n",
       "#OTU ID                                                                     \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...       172.0      9014.0  \n",
       "k__Bacteria;p__Verrucomicrobia;c__Verrucomicrob...     11467.0         0.0  \n",
       "k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__...        46.0        17.0  \n",
       "k__Bacteria;p__Proteobacteria;c__Gammaproteobac...        26.0      9230.0  \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0      1199.0  \n",
       "...                                                        ...         ...  \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0  \n",
       "k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__...         0.0         0.0  \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         0.0  \n",
       "k__Bacteria;p__TM7;c__TM7-3;o__Blgi18;f__;g__              0.0         0.0  \n",
       "k__Bacteria;p__Firmicutes;c__Clostridia;o__Clos...         0.0         2.0  \n",
       "\n",
       "[93 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abundances = pd.read_table(\"exported/genus.tsv\", skiprows=1, index_col=0)\n",
    "abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86543d7e-f24b-4700-8682-cb742183d202",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0568abba-1c85-482c-aefc-c025f616a8eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Image_amplicon_2024.10",
   "language": "python",
   "name": "amplicon_2024.10"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
